### ISSUE TITLE
The WebSocket transport connection timeout does not trigger a reconnection attempt

### ISSUE THREAD

**Issue by yren-ITH** [onsip/SIP.js#706]
**Describe the bug**
The WebSocket transport connection timeout does not trigger a reconnection attempt.
The reconnection attempt is only triggered when the WebSocket itself times out (which can take a very long time, upwards of 1 minute).

**Logs**
https://gist.github.com/yren-ITH/1f39bc32e0d115e30791366cff6ec17e

**To Reproduce (if possible)**
1. Create a UserAgent with a list of dummy wsServers and logging enabled
```
const userAgent = new UA({
  transportOptions: {
    wsServers: {
      `wss:example.com:11111`,
      `wss:example.net:11111`,
      `wss:example.org:11111`
    },
    connectionTimeout: 3,
    maxReconnectionAttempts: 1,
    reconnectionTimeout: 1,
  },
  uri: 'foo@bar.com',
  password: 'foobar',
  log: {
    level: 3,
  },
})
```

**Expected behavior**
* Each connection attempt times out after three seconds.
* One second after the connection times out, a new connection attempt is made with the next WS server in the list.

**Observed behavior**
* Each connection attempt times out after three seconds.
* One minute later, a new connection attempt is made with the next WS server in the list.

**Environment Information**
SIP.js Version 0.14.6
Chrome Version 74.0.3729.169 (Official Build) (64-bit)

**Additional context**
N/A

**Comment by egreenmachine** [comment]
I see the issue. The problem is that I am not sure how to fix it without adding a ton of confusion to the current Transport API. We are going to move the Transport to our new delegate based approach which would allow us to fix this and reduce confusion on the public API front.

**Comment by yren-ITH** [comment]
Could you explain why fixing this would add confusion to the Transport API?
I was assuming that the fix would be as simple as closing the WebSocket inside of the connection timeout.

**Comment by egreenmachine** [comment]
Calling `connect` returns a promise which resolves when the transport is connected and rejects if the transport does not connect. It is currently doing that. The connection fails and the promise rejects. The UA doesn't care about it, but the Transport API is doing the correct thing. If I make it reconnect, it is not clear when the promise should resolve or reject. The easiest thing to do is get rid of the promise and just have the delegate update based on the state of the transport. Then the Transport could do whatever it wants, and update the delegate appropriately.

**Comment by yren-ITH** [comment]
Is there a reason why we would need to change the existing reconnect logic to fix this bug?
A reconnect attempt is already being made in Transport's `onClose` after the initial `connect` promise rejects.

I believe this is how the code currently behaves:
* Connection times out
* connection promise rejects
* WebSocket times out and closes after ~ 1 minute
* `Transport.onClose` is called
* `Transport.reconnect` is called

Closing the WebSocket inside the timeout would result in this behavior:
* Connection times out
* connection promise rejects
* `WebSocket.close` is called
* `Transport.onClose` is called
* `Transport.reconnect` is called

**Comment by egreenmachine** [comment]
The current behavior is definitely broken. I am not arguing with you there. The rejection of the promise is indicating that the application needs to do something. I want to fix it correctly. In the meantime I pushed a fix that cleans up the Websocket when the timeout is hit so that `onClose` and `reconnect` do not get called once we indicate to the application to do something.

**Comment by john-e-riordan** [comment]
The following comment does not speak to what is going on in the current API or provide anything helpful on that front.

But FYI with regard to the new API - a couple of architectural things changed as of 0.15.8...

1. Reconnection logic has been moved to UserAgent
2. A single Transport instance now only deals with a single WebSocket server
3. The UserAgent delegate now provides `onConnect` and `onDisconnect` (which should be used in lieu of the legacy Transport events)
4. One may now call `disconnect()` on a Transport instance before an outstanding call to `connect()` completes in which case `connect()` will reject.
5. Support for configuring multiple servers was removed.

With all that said,  I don't see any technical roadblocks to adding multiple server back in as a feature of the UserAgent (instead of the Transport). Having UserAgent construct and maintain individual Transport instances, and manage the associated fail overs, and swapping them in/out as the Transport being used by the UserAgentCore seems like it would relatively straightforward.

I'm not planning on getting to it in the immediate term, but I'll put it on the 0.16.x roadmap.

**Comment by james-criscuolo** [comment]
This issue has been fixed since 0.15.8, and while there is more pending work for 0.16, the issue itself has been resolved, so I'll close.
